src/_P021_Level.ino

```diff
@@ -18,24 +22,119 @@
+# define P021_DONT_ALWAYS_SAVE      PCONFIG(2)
+# define P021_GUID_DONT_ALWAYS_SAVE "f2"
+
+// Set of individual bits packed in a single PCONFIG(3)
+// Use P021_getConfigBit() and P021_setConfigBit() to access the individual bits
+// For backwards compatibility first bit shall be equal original P021_INVERT_OUTPUT
+# define P021_FLAGS                 PCONFIG(3)
+# define P021_INV_OUTPUT            0 // Invert output signal
+# define P021_GUID_INV_OUTPUT       "b0"
+# define P021_EXT_FUNCT             1 // Extended functionality on setup screen
+# define P021_GUID_EXT_FUNCT        "b1"
+# define P021_INV_INPUT             2 // Invert input comparator
+# define P021_GUID_INV_INPUT        "b2"
+# define P021_SYM_HYSTERESIS        3 // Use symetrical hysteresis
+# define P021_GUID_SYM_HYSTERESIS   "b3"
+# define P021_SLOW_EVAL             4 // 1Hz evaluation i.s.o. 10Hz
+# define P021_GUID_SLOW_EVAL        "b4"
+# define P021_STATE_OUTP            5 // Add state as additional output value
+# define P021_GUID_STATE_OUTP       "b5"
+# define P021_EXTEND_END            6 // Extend state timed from start
+# define P021_GUID_EXTEND_END       "b6"
+
+// Minimum time output shall be active [min]
+# define P021_MIN_TIME              PCONFIG(4)
+# define P021_GUID_MIN_TIME         "f4"
+
+// Maximum time output may be idle [hour]
+# define P021_INTERVAL_TIME         PCONFIG(5)
+# define P021_GUID_INTERVAL_TIME    "f5"
+
+// Duration for forced circulation [min]
+# define P021_FORCE_TIME            PCONFIG(6)
+# define P021_GUID_FORCE_TIME       "f6"
+
+// Operation mode [enum]
+# define P021_OPMODE                PCONFIG(7)
+# define P021_GUID_OPMODE           "f7"
+
+// Setpoint (set level)
+# define P021_SETPOINT              PCONFIG_FLOAT(0)
+# define P021_GUID_SETPOINT         "f8"
+
+// Hysteresis
+# define P021_HYSTERESIS            PCONFIG_FLOAT(1)
+# define P021_GUID_HYSTERESIS       "f9"
+
+// SETP_LAST_STORED is not available on form. Used as storage for autosave function
+# define P021_SETP_LAST_STORED      PCONFIG_FLOAT(2)
+# define P021_GUID_SETP_LAST_STORED "f10"
+# define P021_AUTOSAVE_TIMER        PCONFIG_ULONG(0)
+# define P021_GUID_AUTOSAVE_TIMER   "f11"
+
+// Positions in userVar array for the output values of this plugin
+// For now we only advertise output & state. Autosave bookkeeping is hidden and used as static storage
+#define P021_VALUE_OUTPUT           0 // Switch output, logical state [inactive/active]
+#define P021_VALUE_STATE            1 // Control state, see P021_control_state
+#define P021_VALUE_AUTOSAVE_FLAG    2 // Autosave bookkeeping
+#define P021_VALUE_AUTOSAVE_TIME    3 // Autosave bookkeeping
+
+// Make time interpretation easier by converting millis to the longer periods we want to check
+#define millis2seconds(x) ((x) / ((ulong)1000))
+#define millis2minutes(x) ((x) / ((ulong)1000 * 60))
+#define millis2hours(x) ((x) / ((ulong)1000 * 60 * 24))
```

Comment:
First question: Why???

Second question:
Please take another look at the code ;)

Reply:
Also these are integer operations and thus it will always round down.

For example: 59999 msec will be rounded to 0 when calling the `millis2minutes` macro.
Is that what you intend to happen?

Reply:
At several places the configuration parameters are set in seconds/minutes/hours. Entered time values are used as a minimum value for an interval. Or phrased different compare if the interval is above the rounded up value. I don't like repeating the same conversion so I made them macros. In hind sight it is more reliable to compare in the millisecond domain. So defining minutes2millis and use it at the other side of the compare..

Reply:
Reworked way of handling time values. Now delays are stored in PCONFIG_ULONG as seconds. For user interface convert these for convenient human interpretation (two scales will be provided). For the algorithm the seconds are compared to millis(). The time scaling is done by reworked macros.